rust/nft-candy-machine/src/lib.rs

```diff
@@ -286,7 +300,8 @@ pub mod nft_candy_machine {
         let account = config.to_account_info();
         let current_count = get_config_count(&account.data.borrow())?;
         let mut data = account.data.borrow_mut();
-
+        let is_new_configuration = data.len() >= CONFIG_ARRAY_START + 4 + (current_count as usize) * CONFIG_LINE_SIZE + 5;
+        msg!("is_new_configuration: {} for {} and {} lines", is_new_configuration, data.len(), current_count);
```

Comment:
Yeah the more I look at this, the more I dislike it. I talked with @armaniferrante about how to do this the right way and he came up with this good example:

```rust
#[account]
pub struct MyAccountV1 {...}

#[account]
pub struct MyAccountV2 {...}

#[derive(Clone)]
pub enum MyAccount {
   V1(MyAccountV1),
   V2(MyAccountV2),
}

impl AccountDeserialize for MyAccount {
  fn try_deserialize(buf: &mut &[u8]) -> Result<Self, ProgramError> { 
    if discriminator(buf) == MyAccountV1::discriminator() {
       Ok(MyAccount::V1(MyAccountV1::try_deserialize(buf)))
    } else if discriminator(buf) == MyAccountV2::discriminator() {
       Ok(MyAccount::V2(MyAccountV2::try_deserialize(buf)))
    } else {
       Err(...)
     }
  }
  fn try_deserialize_unchecked(buf: &mut &[u8]) -> Result<Self, ProgramError> {
    // similar to above
  }
}

impl AccountSerialize for MyAccount {...}
```

Doing this would allow us to essentially turn the Config into an enum of ConfigV1 and ConfigV2. I think this is how it has to be done. It's more work but if we make more changes in the future it'll be much more extensible than this hacky way.

Reply:
Cannot agree more, it's way more elegant, and less error prone, than what we ended up with. I'll try to implement it this way asap! 

Thanks for pointing out this solution.

Reply:
I liked this idea, but unfortunately, if I understood everything well, I think it might hit some limitations in the anchor library that may need to be worked around.

First, we cannot use [tuple enum variants](https://github.com/project-serum/anchor/blob/90df0b19766ef2a0937312a3864f44e49bb16cad/ts/src/coder/idl.ts#L131) for the `Config` enum for now, but it can still be serialized properly by anchor if we define named variants, which ends with the following in the IDL:

```
{
  "name": "Config",
  "type": {
    "kind": "enum",
    "variants": [
      {
        "name": "V1",
        "fields": [
          {
            "name": "config",
            "type": {
              "defined": "ConfigV1"
            }
          }
        ]
      },
      {
        "name": "V2",
        "fields": [
          {
            "name": "config",
            "type": {
              "defined": "ConfigV2"
            }
          }
        ]
      }
    ]
  }
}
```

Unfortunately, it seems anchor is not serializing the related `ConfigV1` and `ConfigV2` types in the generated IDL, and the following error pops client-side when instantiating the program stub: 

```
IdlError: Type not found: {"name":"config","type":{"defined":"ConfigV1"}}
```

I have to check why in the anchor library and maybe ask them about that, but to make a long story short, I cannot guarantee that I will end up with a working implementation this way, but I'll update this PR if I do!

I kept both working implementations to spare branches in my fork, don't hesitate to ask me if you want me to create a PR for one of them. One of them used the `u32::MAX` (or `0`, which is also unused) values as trigger in the `max_number_of_lines` field of the configuration, and the other is this PR, which serializes this trigger as a boolean in the custom borsh data of the configuration in a backward compatible way.

Without considering this limitation, after trying different ways, my feeling is that the `max_number_of_lines` approach was less error prone than the custom borsh serialization, and that it is more compact and readable than a custom enumeration, even if it is less elegant and extensible. But this can definitely be the consequence of my rudimentary knowledge of the solana ecosystem.

In the between, people can still use a custom program using token metadatas, or a forked version of the candy machine, if they need to emit large collectible without storing every single configuration line on the blockchain.

It was interesting to work on this and to try these different approaches, thank you again for your insights on this! As I said, I'll update the PR if I find enough spare time to dig this serialization issue in anchor, and I'll keep an eye on this repo to see if this template feature is implemented somehow in the future!

Cheers!